src/System.Private.CoreLib/shared/System/Exception.cs

```diff
@@ -120,26 +120,48 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
 
         public override string ToString()
         {
-            string s = GetClassName();
+            // Get the lengths of the StackTrace and _innerException first so we know how large to make the ValueStringBuilder
+            string? stackTrace = StackTrace;
+            string? innerException = _innerException?.ToString();
+
+            int initialLength = 128; // arbitrary initial guess
+            if (stackTrace != null)
+            {
+                initialLength += stackTrace.Length;
+            }
+            if (innerException != null)
+            {
+                initialLength += innerException.Length;
+            }
+
+            var sb = new ValueStringBuilder(initialLength);
```

Comment:
This usage makes me a bit nervous. I can easily imagine a case where an exception is thrown because something has gone wrong with the internal shared `ArrayPool` instance, which would mean that we can't reliably call `Exception.ToString` to get the human-readable exception message. Is optimizing this particular method of high importance?

Reply:
Yes /cc @NickCraver

Reply:
Yes; a lot of optimizations have made to make this path faster and less allocatey https://github.com/dotnet/coreclr/pull/20448 https://github.com/dotnet/coreclr/pull/20463 https://github.com/dotnet/coreclr/pull/20779 https://github.com/dotnet/coreclr/pull/20795 https://github.com/dotnet/coreclr/pull/21462 /cc @NickCraver

Reply:
Yes, but most of those changes were general goodness and did not come with trace-offs.

Once there are trade-offs, the GC allocations are secondary metric. Not something to optimize for as primary goal.

Reply:
Is it any different then getting Dictionary wrong when you can't throw exceptions because it can't look up the resource strings?

Reply:
The difference is that ArrayPool is unsafe construct that is easy to use incorrectly. If the diagnostics for that incorrect use is based on the ArrayPool again, the original error may be much harder to diagnose.

Reply:
I can change this to use StringBuilder instead.

What about the case of StackTrace, which is used by Exception.ToString().  Presumably your concerns apply there, too?

Reply:
I suppose this same concern would apply to anything diagnostics related, but this isn't an area I'm intimately familiar with. That could lead me to make incorrect trade-off decisions.

Reply:
As an addendum to my comment, I'm mindful that we don't want to go down a rabbit hole. For example, since `StringBuilder` now uses `ArrayPool` under the covers, do we want to avoid `StringBuilder` usage in the `Exception.ToString` method? Eventually this turns into auditing every single line of code that `Exception.ToString` might call, which might not be a worthwhile exercise. Hence why I'm not a good authority to speak on this issue. :)

Reply:
I'm changing Exception.ToString to use String.Create. I'll leave StackTrace as is for now.

Reply:
> StringBuilder now uses ArrayPool under the covers,

It does not.

Reply:
> It does not.

Derp, sorry. I was confusing `AllocUninitializedArray` with `ArrayPool`. Corrected my original comment.

Reply:
> Once there are trade-offs, the GC allocations are secondary metric. Not something to optimize for as primary goal.

I feel a bit guilty for increasing allocations from making async stack traces more intelligible due to the increased use of reflection

Reply:
(allocations from checking method attributes)